lib/meadow/data/csv_export.ex

```diff
@@ -0,0 +1,86 @@
+defmodule Meadow.Data.CsvExport do
+  @moduledoc """
+  Generates a CSV representation of works matching an Elasticsearch query
+  """
+
+  alias Meadow.Data.Schemas.{WorkAdministrativeMetadata, WorkDescriptiveMetadata}
+  alias Meadow.Utils.ElasticsearchResultStream
+  alias NimbleCSV.RFC4180, as: CSV
+
+  @line_separator "\r\n"
+
+  @doc """
+  Generates the CSV output for an Elasticsearch query
+
+  Example:
+    iex> generate_csv(%{query: %{match_all: %{}}})
+    "{\"query\":{\"match_all\":{}}}\r\nid,accession_number..."
+  """
+  def generate_csv(query) when is_binary(query) do
+    [query, @line_separator, generate_header(), generate_rows(query)]
+    |> IO.iodata_to_binary()
+  end
+
+  def generate_csv(query), do: Jason.encode!(query) |> generate_csv()
+
+  defp fields do
+    [["id"], ["accessionNumber"]] ++
+      fields_for(WorkAdministrativeMetadata, "administrativeMetadata") ++
+      fields_for(WorkDescriptiveMetadata, "descriptiveMetadata")
+  end
+
+  defp fields_for(module, prefix) do
+    apply(module, :field_names, [])
+    |> Enum.map(fn field -> [prefix, Inflex.camelize(field, :lower)] end)
+  end
+
+  defp generate_header do
+    [
+      fields()
+      |> Enum.map(fn field_path ->
+        field_path |> List.last() |> Inflex.underscore()
+      end)
+    ]
+    |> CSV.dump_to_iodata()
+  end
+
+  defp generate_rows(query) do
+    query
+    |> ElasticsearchResultStream.results()
+    |> Stream.map(&to_row/1)
+    |> CSV.dump_to_iodata()
+  end
+
+  defp to_row(hit) do
+    fields()
+    |> Enum.map(fn field_path ->
+      hit
+      |> get_in(["_source" | field_path])
+      |> to_field()
+    end)
+  end
+
+  defp to_field(value) when is_list(value) do
+    value
+    |> Enum.map(&to_field/1)
+    |> Enum.join(" | ")
+  end
+
+  defp to_field(%{"role" => role, "term" => %{"label" => label}}) do
+    case normalize_role(role) do
+      nil -> label
+      prefix -> [prefix, label] |> Enum.join(":")
```

Comment:
@mbklein I know this is a bit ahead of scope, but if this is the human readable csv, we should format like "Klein, Michael (Developer)" rather than "dev:Klein, Michael". A use case will be to use this data in other systems, and I don't think the prefixes will make much sense. As I'm writing this, maybe it makes sense to have a future feature that allows a "format" flag.

Reply:
Oh, OK. Cool. That makes a lot of sense.

Reply:
Wait. @davidschober - I thought we weren't doing the human readable csv? We're just doing the export/reimport?

Reply:
@kdid is right. I implemented this incorrectly from the get-go. Fortunately, the code is flexible enough to make it an easy fix. 😄